DirectProgramming/DPC++/GraphTraversal/hidden-markov-models/src/hidden_markov_models.cpp

```diff
@@ -0,0 +1,193 @@
+//==============================================================
+// Copyright © Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+//
+// Hidden Markov Models: this code sample implements the Viterbi algorithm which is a dynamic 
+// programming algorithm for findingthe most likely sequence of hidden states—
+// called the Viterbi path—that results in a sequence of observed events,
+// especially in the context of Markov information sources and HMM.
+//
+// - Initially, the dataset for algorithm processing is generated : initial states probability 
+// distribution Pi, transition matrix A, emission matrix Band the sequence or the observations 
+// produced by hidden Markov process.
+// - First, the matrix of Viterbi values on the first states are initialized using distribution Pi 
+// and emission matrix B.The matrix of back pointers is initialized with default values - 1.
+// - Then, for each time step the Viterbi matrix is set to the maximal possible value using A, B and Pi.
+// - Finally, the state with maximum Viterbi value on the last step is set as a final state of 
+// the Viterbi pathand the previous nodes of this path are detemined using the correspondent rows 
+// of back pointers matrix for each of the steps except the last one.
+//
+// Note: The implementation uses logarithms of the probabilities to process small numbers correctly
+// and to replace multiplication operations with addition operations.
+
+#include <CL/sycl.hpp>
+#include <iostream>
+#include <limits>
+#include <math.h>
+#include <iostream>
+#include <cstdio>
+#include "dpc_common.hpp"
+
+using namespace sycl;
+using namespace std;
+
+// Matrix size constants.
+// The number of hidden states N.
+constexpr int N = 20;
+// The number of possible observations M.
+constexpr int M = 20;
+// The lenght of the hidden states sequence T.
+constexpr int T = 20;
+// The parameter for generating the sequence.
+constexpr int seed = 0;
+// Minimal double to initialize  logarithms for Viterbi values equal to 0.
+constexpr double MIN_DOUBLE = -1.0 * std::numeric_limits<double>::max();
+
+bool ViterbiCondition(double x, double y, double z, double compare);
+
+int main() {
+    // Initializing and generating initial probabilities for the hidden states.
+    double(*pi) = new double[N];
+    for (int i = 0; i < N; ++i) {
+        pi[i] = sycl::log10(1.0f / N);
+    }
+
+    // Сatch asynchronous exceptions.
+    auto exception_handler = [] (sycl::exception_list exceptions) {
+        for (std::exception_ptr const& e : exceptions) {
+            try {
+                std::rethrow_exception(e);
+            } catch(sycl::exception const& e) {
+                std::cout << "Caught asynchronous SYCL exception:\n"
+                << e.what() << std::endl;
+            }
+        }
+    };
+    
+    try {
+        //Device initialization.
+        queue q(default_selector{}, exception_handler);
```

Comment:
pass the exception_handler as below
queue q(default_selector{}, dpc_common::exception_handler);